File: Find the Length of the Longest Common Prefix.py
class Node:
    def __init__(self):
        self.route = {}
        self.end = False

    def add(self, word, idx):
        if len(word) == idx:
            self.end = True
            return

        c = word[idx]
        if c not in self.route:
            self.route[c] = Node()
        self.route[c].add(word, idx+1)

    def find(self, word, idx):
        if len(word) == idx:
            return len(word)

        c = word[idx]
        if c not in self.route:
            return idx
        return self.route[c].find(word, idx+1)

class Solution:
    def longestCommonPrefix(self, arr1: List[int], arr2: List[int]) -> int:
        trie = Node()
        for word in arr1:
            word = str(word)
            trie.add(word, 0)

        maxx = 0
        for word in arr2:
            word = str(word)
            maxx = max(maxx, trie.find(word, 0))
        return maxx

File: Find the Length of the Longest Common Prefix.cpp
class Node {
public:
    unordered_map<char, Node> route;
    bool end = false;
    Node() {};
    void add(string& word, int idx) {
        if (word.size() == idx) {
            end = true;
            return;
        }

        auto c = word[idx];
        if (!route.contains(c)) route[c] = Node();
        route[c].add(word, idx+1);
    }

    int find(string& word, int idx) {
        if (word.size() == idx) return word.size();

        auto c = word[idx];
        if (!route.contains(c)) return idx;
        return route[c].find(word, idx+1);
    }
};

class Solution {
public:
    int longestCommonPrefix(vector<int>& arr1, vector<int>& arr2) {
        auto trie = Node();
        for (int i: arr1) {
            auto word = to_string(i);
            trie.add(word, 0);
        }

        int maxx = 0;
        for (int i: arr2) {
            auto word = to_string(i);
            maxx = max(maxx, trie.find(word, 0));
        }

        return maxx;
    }
};
